gimie/sources/gitlab.py

```diff
@@ -1,14 +1,287 @@
+        # the code below does not work, returns - if you have permission- the GitLab specific licence
+        # resp = requests.get(url=f"{GL_API_rest}/license/{self.identifier}")
+        # if resp.status_code == 200:
+        #     self.license = resp.json()
+
+    def _fetch_repo_data(self, url: str) -> Dict[str, Any]:
+        """Fetch repository metadata from GraphQL endpoint."""
+        response = query_repo_graphql(url, self._set_auth())
+        if "errors" in response:
+            raise ValueError(response["errors"])
+
+        return response["data"]["project"]
+
+    def _set_auth(self) -> Any:
+        """Set authentication headers for Gitlab API requests."""
+        try:
+            if not self.Gitlab_token:
+                self.Gitlab_token = os.environ.get("ACCESS_TOKEN")
+                assert self.Gitlab_token
+            headers = {"Authorization": f"token {self.Gitlab_token}"}
+
+            login = requests.get(f"{GL_API}/user", headers=headers)
+            assert login.json().get("login")
+        except AssertionError:
+            return {}
+        else:
+            return headers
+
+    def _get_author(self, node: Dict[str, Any]) -> Union[Organization, Person]:
+        """Given the GraphQL node for a repository owner,
+        return the author as a Person or Organization object."""
+        # Is this the best test?
+        if "username" in node:
+            return self._get_user(node)
+        return self._get_organization(node)
+
+    def _get_organization(self, node: Dict[str, Any]) -> Organization:
+        """Extract details from a GraphQL organization node."""
+        return Organization(
+            _id=node["webUrl"],
+            name=node["name"],
+            description=node["description"],
+            logo=node["avatarUrl"],
+        )
+
+    def _get_user(self, node: Dict[str, Any]) -> Person:
+        """Extract details from a GraphQL user node."""
+        # Get user's affiliations
+        # To Do
+        return Person(
+            _id=node["webUrl"],
+            identifier=node["username"],
+            name=node["name"],
+            email=node["publicEmail"],
+        )
+
+
+class GitlabExtractorSchema(JsonLDSchema):
+    """This defines the schema used for json-ld serialization."""
+
+    _id = fields.Id()
+    name = fields.String(SDO.name)
+    identifier = fields.String(SDO.identifier)
+    sourceOrganization = fields.Nested(SDO.isPartOf, OrganizationSchema)
+    author = fields.Nested(
+        SDO.author, [PersonSchema, OrganizationSchema], many=True
+    )
+    contributors = fields.Nested(SDO.contributor, PersonSchema, many=True)
+    prog_langs = fields.List(SDO.programmingLanguage, fields.String)
+    download_url = fields.Raw(SDO.downloadUrl)
+    description = fields.String(SDO.description)
+    date_created = fields.Date(SDO.dateCreated)
+    date_modified = fields.Date(SDO.dateModified)
+    # license = IRI(SDO.license)
```

Comment:
schema:CodeRepository is the (resolvable) URL of the repo, which should be path in this case. This could be useful if we end up using something different for the instance URI.

Reply:
not sure what you are referring to here: can you expand? or we can quickly discuss over zoom

Reply:
I was referring to your comment in the line just below:
```
# what is this ?
# path = IRI(SDO.CodeRepository)
```
i.e., I think you don't need to comment out the line `path = IRI(SDO.CodeRepository)`